doc/man1/mql.1
.\" Man page for mql
.\" Use the following command to view man page:
.\"
.\"  tbl mql.1 | nroff -man | less
.\"
.TH MQL 1 "August 14, 2024"
.SH NAME
mql \- A front-end to the Emdros engine using the MQL query language
.SH SYNOPSIS
\fBmql\fR [ options ] filename
.br
\fBmql\fR [ options ] 
.br
.SH DESCRIPTION
\fBmql\fR is a front-end to Emdros using the MQL query language.  If a
filename is given on the command-line, the contents of that file are
piped through the MQL engine.  If no filename is given on the
command-line, stdin is used to get the MQL input.  The output is given
on stdout, or on stderr if unsuccessful.

.SH OPTIONS
\fBmql\fR supports the following command-line switches:
.TP 6
.BI \-\-help
show help
.TP
.BI \-V\ ,\ \-\-version
show version
.TP
.BI \-\-xml\ 
output-format is XML
.TP
.BI \-\-cxml\ 
output-format is compact XML
.TP
.BI \-\-json\ 
output-format is pretty-printed JSON
.TP
.BI \-\-cjson\ 
output-format is compact JSON
.TP
.BI \-\-console
output-format is console (this is the default)
.TP
.BI \-\-incomplete-output
the caller is prepared to expect incomplete output.  The tradeoff is
that, with this option, we gain some speed due to flushing the output
buffer all the time being unnecessary.
.TP
.BI \-o\ ,\ \-\-output output_filename
print the output on this file. If '-', is given, use stdout.
.TP
.BI \-d\ ,\ \-\-dbname \ dbname 
set initial dbname (default is 'emdf')
.TP
.BI \-h\ ,\ \-\-host \ hostname 
set db back-end hostname to connect to (default is 'localhost').  Has
no effect on SQLite.
.TP
.BI \-u\ ,\ \-\-user \ user
set database user to connect as (default is 'emdf').  Has no effect on
SQLite.
.TP
.BI \-p\ ,\ \-\-password \ password
set password to use for the database user.  Has no effect on SQLite,
unless you have an encryption-enabled SQLite, in which case this gets
passed as the key.
.TP
.BI \-b\ ,\ \-\-backend \ backend
set database backend to `backend'. Valid values are: For PostgreSQL:
"p", "pg", "postgres", and "postgresql". For MySQL: "m", "my", and
"mysql". For SQLite 2.X.X: "2", "s", "l", "lt", "sqlite", and
"sqlite2". For SQLite 3.X.X: "3", "s3", "lt3", and "sqlite3".
.TP
.BI \-n\ ,\ \-\-nop 
do not print results (no output)


.SH RETURN VALUES
.TP
.BR 0 " Success"
.TP
.BR 1 " Wrong usage"
.TP
.BR 2 " Connection to backend server could not be established"
.TP
.BR 3 " An exception occurred (the type is printed on stderr)"
.TP
.BR 4 " Could not open file"
.TP
.BR 5 " Database error"
.TP
.BR 6 " Compiler error (error in MQL input)"
.SH AUTHORS
Written Ulrik Sandborg-Petersen (ulrikp@emdros.org).


NEWS
This file is for release-notes for Emdros.

- *** Version 3.9.1 *** 

Release notes for Emdros 3.9.1
==============================

- Emdros Query Tool (wxWidgets) / eqtu(1): Add ability to run query,
  emit as HTML, and exit immediately. Add command line switches --html
  and -o outfilename to implement this.

- eqtc(1): Add -o, --output outfilename switch.

- *** Version 3.9.0 *** 

Release notes for Emdros 3.9.0
==============================

- Update SQLite to version 3.43.1.

- Build system:

  - Enable compilation on macOS Sonoma.

  - Fix the Programmer's Reference guide's build system in various ways.

  - Fix Python 3 compatibility in the build system.

  - Fix various issues with the Debian .deb build system.

  - build-deb.sh: Add the possibility to set the shell environment
    variable EMDROS_CONFIGURE_SWITCHES which will pass switches to the
    Emdros configure script when building the .deb.

  - Bump autoconf AC_PREREQ from 2.69 to 2.71.

  - Update the install_mac.sh.in script, so it actually gives an error
    message and stops if an error is encountered. Also, copy the
    documentation as part of the macOS install.

- Changes for the JSON classes:

  - Add new method JSONValue::castAsCompactString(bool bEscapeAsUnicode
    = false);

  - Document the JSON functionality in the Programmer's Reference
    Guide, and inline in the code.

  - Add parameter bool bEscapeAsUnicode, toJSONValue::castAsString()
    with default value 'false', for backwards compatiblity.

  - Add new method JSONValue::castAsCompactString(bool
    bEscapeAsUnicode = false).

  - Fix bugs in:
  
    - JSONValue::deleteList(void)
    - JSONValue::deleteObject(void)

    Bugfix: Only delete the list / object if the kind is actually
    kJSONList or kJSONObject respectively.

- SWIG bindings changes:

  - Extend the example for the Node.js SWIG bindings to exercise the
    renderobjects machinery.

  - Add support for C++ STL <map> typemaps in various interfaces.

- Emdros Rest Wrapper (util/erwc.pp, include/erw.h):

   - Move it closer towards usability.

- Emdros Query Tool:

   - Fix the connection dialog so it works again, now being able to
     validate the backend chooser again.



- *** Version 3.8.0 *** 

Release notes for Emdros 3.8.0
==============================

- Remove support for PHP5 bindings.

- Add support for PHP8 bindings

- Add experimental support for Node.js bindings.

- Upgraded SQLite3, PCRE, and ZLib to latest versions as of 2022-12-31.

- Test suites extended

- Assorted bugfixes

- Fixes to the changing sands of compilers, languages, platforms,
  security settings, operating systems... Sisyphos certainly has his
  work rolled out for him, and he gives thanks for having a job...

- MySQL is currently not supported on Windows, or on Mac.

- PostgreSQL is currently not supported on Windows, or on Mac.

- SWIG bindings are currently not supported on Windows.




- *** Version 3.7.0 *** 

Release notes for Emdros 3.7.0
==============================

- Emdros has been relicensed under the MIT license.

- All copyright in Emdros and related documentation and build scripts
  has been assigned to Ulrik Sandborg-Petersen's holding company,
  Sandborg-Petersen Holding ApS.  This is so as to be able to serve
  customers better with support and other Emdros-related help.

- A new construct was added to the topographic part of MQL: computed
  features, which take a monad set and returns a monad or integer:

  - first_monad(<monad_set_name>) returns the smallest (first) monad.
  - last_monad(<monad_set_name>) returns the largest (last) monad.
  - monad_count(<monad_set_mame>) returns the cardinality.
  - monad_set_length(<monad_set_mame>) returns last_monad - first_monad + 1.

  If the monad_set_name (and the parentheses) are missing, then the
  privileged monad set "monads" is assumed.

  For example:

  SELECT ALL OBJECTS
  WHERE
  [Clause
     monad_set_length > 15 AND monad_count <= 15
  ]

  This also works with object references:

  // Finds sentences and clauses which are co-extensive.
  SELECT ALL OBJECTS
  WHERE
  [Sentence AS s1
     [Clause monad_count = s1.monad_count]
  ]

- The implementation of NOTEXIST object blocks has been corrected, so
  that it conforms to the documentation in the "MQL Query Guide".
  Specifically, when you have two or more consecutive NOTEXIST object
  blocks at the beginning of a block_string, the results are now as
  described in the MQL Query Guide.  Specifically:

  a) The order of the NOTEXIST object blocks no longer matters (it did
  before), and

  b) The scope of all of the consecutive NOTEXIST object blocks is now
  the entire Universe and the entire Substrate.

  Thanks to Camil Staps for bringing this to Ulrik's attention.

- The implementation of NOTEXIST is now much, much faster.

- Regular expressions are now treated as UTF8, not as legacy-encoding
  raw 8-bit data.  This is true for MQL, as well as for the two kinds
  of configuration files (with and without [ini-file-like] sections.

  In particular, character classes can now be used correctly with
  UTF8-encoded characters.  For example, the following now does the
  Correct Thing(TM):

  SELECT ALL OBJECTS WHERE [Word surface="^.*[ÆØÅæøå].*$".

  Before, this would not work, since the UTF-8 encoded characters
  would be treated, not as distinct characters, but as the raw 8-bit
  bytes of which they were made up.

- The Harvest library template language has some new templates:

  {{ setvar 'output-var-name' substring var 'input-var-name' from /startindex/ maxlength /max-length/ }}
  {{ setvar 'output-var-name' substring attrib 'input-attrib-name' from /startindex/ maxlength /max-length/ }}

src/mql.cpp
/*
 * mql.cpp
 *
 * A front-end to the MQL parser
 * Created: 5/1-2001 (1st of May, 2001)
 * Last update: 8/14-2024
 *
 * Return codes:
 * 
 * 0 : Success
 * 1 : Wrong usage
 * 2 : Connection to backend server could not be established
 * 3 : An exception occurred
 * 4 : Could not open file
 * 5 : Database error
 * 6 : Compiler error
 *
 */

/*
 * Copyright (C) 2001-2018     Ulrik Sandborg-Petersen
 * Copyright (C) 2018-present  Sandborg-Petersen Holding ApS, Denmark
 *
 * Licensed under the MIT License.
 *
 * Please see the file COPYING in the root of the sources for more details.
 *
 */



#include <emdros-config.h>
// Do this because emdros-config might #define malloc,
// in which case some systems will fail to use the real 
// malloc, thus causing build errors.
#undef malloc


#include <monads.h>
#include <iostream>
#include <sstream>
#include <fstream>
#include <string_func.h>
#include <emdfdb.h>
#include <table.h>
#include <emdros-lconfig.h>
#include <emdf_output.h>
#include <mql_execute.h>
#include <mql_sheaf.h>
#include <emdros_environment.h>
#include <encryption.h>
#include <opt.h>
#include <messages.h>



int exec_file(EmdrosEnv *pEE, std::string filename, bool bPrintOutput, bool bCallerAcceptsIncompleteOutput, MQLResultCallback *pCallback)
{
	UNUSED(bCallerAcceptsIncompleteOutput);
	
	bool bResult;
	std::string strError;
	std::ifstream fin;
	int nResult;
	fin.open(filename.c_str());
	if (!fin) {
		std::cerr << "FAILURE: Could not open file " << filename << " for reading." << std::endl;
		std::cerr << "Action aborted." << std::endl;
		nResult = 4;
	}
	if (!pEE->executeStream(fin, bResult, bPrintOutput, true, pCallback)) {
		std::cerr << "FAILURE: Database error executing " << filename << std::endl;
		nResult = 5;
	} else {
		if (!bResult) {
			// std::cerr << "FAILURE: Compiler error executing " << filename << std::endl;
			nResult = 6;
		} else { 
			// std::cout << "SUCCESS executing file " << filename << std::endl;
			nResult = 0;
		}
	}

	// Return result
	return nResult;
}


int exec_cin(EmdrosEnv *pEE, bool bPrintOutput, bool bCallerAcceptsIncompleteOutput, MQLResultCallback *pCallback)
{
	UNUSED(bCallerAcceptsIncompleteOutput);
	
	bool bResult;
	std::string strError;
	int nResult;
	if (!pEE->executeStream(std::cin, bResult, bPrintOutput, true, pCallback)) {
		std::cerr << "FAILURE: Database error executing stdin." << std::endl;
		nResult = 5;
	} else {
		if (!bResult) {
			// std::cerr << "FAILURE: Compiler error executing stdin." << std::endl;
			nResult = 6;
		} else { 
			// std::cout << "SUCCESS executing stdin." << std::endl;
			nResult = 0;
		}
	}

	// Return result
	return nResult;
}

void show_version(std::ostream& ostr)
{
	ostr << "mql from Emdros version " << EMDROS_VERSION << std::endl;
	printBackendsAvailable(ostr);
}

void print_usage(std::ostream& ostr)
{
	show_version(ostr);
	ostr << "Usage:" << std::endl;
	ostr << "   mql [options] [filename] " << std::endl;
	ostr << "       If no filename is given, use stdin." << std::endl;
	ostr << "   mql (--help|-V|--version)" << std::endl;
	ostr << "OPTIONS:" << std::endl;
	ostr << "   --xml                Output XML format" << std::endl;
	ostr << "   --cxml               Output compact XML format" << std::endl;
	ostr << "   --json               Output JSON format" << std::endl;
	ostr << "   --cjson              Output compact JSON format" << std::endl;
	ostr << "   --console            Output console-friendly format" << std::endl;
	ostr << "   --incomplete-output  Caller accepts incomplete output while processing" << std::endl;
	ostr << "   -o , --output fn     Output to this filename. '-' means stdout." << std::endl;
	printUsageStandardArguments(ostr);
	ostr << "   -d , --dbname db     Set initial database to db" << std::endl;
	ostr << "   -n , --nop           Do not print results (no output)" << std::endl;
	ostr << "DEFAULTS:" << std::endl;
	ostr << "   --console" << std::endl;
	ostr << "   -d emdf" << std::endl;
	ostr << "   -o -" << std::endl;
	printUsageDefaultsOfStandardArguments(ostr);
}

int main(int argc, char* argv[])
{
	// Set defaults
	eOutputKind output_kind = kOKConsole;
	bool bCallerAcceptsIncompleteOutput = false;
	std::string initial_db;
	std::string filename;
	std::string hostname;
	std::string user;
	bool bPrintOutput = true;
	std::string password;
	eBackendKind backend_kind;
	eCharsets charset;

	addStandardArguments(true // Do add -e | --encoding
			     );

	addOption("--xml", "--xml");
	addOption("--cxml", "--cxml");
	addOption("--json", "--json");
	addOption("--cjson", "--cjson");
	addOption("--console", "--console");
	addOption("--incomplete-output", "--incomplete-output");
	addOption("-n", "--nop");


	addOption("-d", "--dbname", true, 
		  "emdf",
		  "ERROR: -d and --dbname must have a database-name as their argument.\n"
		  "       example: -d book1\n");

	addOption("-o", "--output", true, 
		  "-",
		  "ERROR: -o and --output must have a filename as their argument.\n"
		  "       example: -o output.sheaf.txt\n");

	// Parse arguments
	std::list<std::string> surplus_arguments;

	std::string output_filename;

	std::string error_message;
	if (!parseArguments(argc, argv, error_message, surplus_arguments)) {
		print_usage(std::cerr);
		std::cerr << error_message << std::endl;
		return 1;
	} else {
		bool bShowVersion;
		bool bShowHelp;
		std::string error_message;
		if (!getStandardArguments(bShowVersion, bShowHelp,
					  hostname,
					  user,
					  password,
					  backend_kind,
					  charset,
					  error_message)) {
			print_usage(std::cerr);
			std::cerr << error_message << std::endl;
			return 1;
		}

		// -V | --version
		if (bShowVersion) {
			show_version(std::cout);
			return 0;
		}

		// --help
		if (bShowHelp) {
			print_usage(std::cout);
			return 0;
		}

		// Output kind
		if (getArgumentPresent("--xml")) {
			output_kind = kOKXML;
		}
		if (getArgumentPresent("--cxml")) {
			output_kind = kOKCompactXML;
		} 
		if (getArgumentPresent("--json")) {
			output_kind = kOKJSON;
		}
		if (getArgumentPresent("--cjson")) {
			output_kind = kOKCompactJSON;
		} 
		if (getArgumentPresent("--console")) {
			output_kind = kOKConsole;
		}
		if (getArgumentPresent("--incomplete-output")) {
			bCallerAcceptsIncompleteOutput = true;
		}

		// --nop | -n
		if (getArgumentPresent("-n")) {
			bPrintOutput = false;
		}

		if (getArgumentPresent("-o")) {
			getArgumentValue("-o", output_filename);
		} else {
			// "-" == stdout
			output_filename = "-";
		}

		getArgumentValue("-d", initial_db);

		if (surplus_arguments.size() == 0) {
			filename = "";
		} else if (surplus_arguments.size() == 1) {
			filename = *surplus_arguments.begin();
		} else {
			std::cerr << "Error: You can only specify one filename. You specified: '" << joinList("' and '", surplus_arguments) << "'." << std::endl;
			return 1;
		}
	}

	int nResult = 0;
	EmdrosEnv *pEE = 0;
	MQLResultCallbackOutput *pCallback = 0;
	std::ostream *pOut = 0;
	
	try {
		if (output_filename == "-") {
			// Make connection, print on stdout
			pEE = new EmdrosEnv(output_kind, 
					    charset,
					    hostname,
					    user, password,
					    initial_db,
					    backend_kind);
		} else {
			pOut = new std::ofstream(output_filename.c_str());

			if (!*pOut) {
				std::cerr << "FAILURE: Could not open file " << output_filename << " for writing." << std::endl;
				std::cerr << "Action aborted." << std::endl;
				nResult = 4;
			} else {
				pEE = new EmdrosEnv(pOut,
						    output_kind, 
						    charset,
						    hostname,
						    user, password,
						    initial_db,
						    backend_kind);
			}
		}

		if (nResult == 0) {
			// Zero-fill password
			zeroFillString(password);
		
			// Check that we connected
			if (!pEE->connectionOk()) {
				emdrosMessageConnectionNotOK(&std::cerr, backend_kind);
				delete pEE;
				return 2;
			}
			
			bool bFlushAfterEveryResult = !bCallerAcceptsIncompleteOutput;
			if (bPrintOutput) {
				pCallback = new MQLResultCallbackOutput(pEE->getMQLEE(), bFlushAfterEveryResult);
			} else {
				pCallback = 0;
			}
			
			if (filename == "") {
				nResult = exec_cin(pEE, bPrintOutput, bCallerAcceptsIncompleteOutput, pCallback);
			} else {
				nResult = exec_file(pEE, filename, bPrintOutput, bCallerAcceptsIncompleteOutput, pCallback);
			}
		}
	} catch (EMdFNULLValueException& e) {
		std::cerr << "ERROR: EMdFNULLValueException (Database error)..." << std::endl;
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBDBError& e) {
		std::cerr << "ERROR: EMdFDBDBError (Database error)..." << std::endl;
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBException& e) {
		std::cerr << "ERROR: EMdFDBException (Database error)..." << std::endl;
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableColumnException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableRowException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (BadMonadsException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (WrongCharacterSetException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFOutputException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EmdrosException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (...) {
		std::cerr << "Unknown exception occurred.  Program aborted." << std::endl;
	} 

	// Clean up
	if (pOut != 0) {
		// Also closes the file
		delete pOut;
	}
	
	delete pEE;
	delete pCallback;


	return nResult;
}


